fallback/s2a_fallback.go

```diff
@@ -0,0 +1,150 @@
+		fallbackTLSConfig := tls.Config{
+			ServerName: fallbackServerName,
+			NextProtos: []string{alpnProtoStrH2},
+		}
+		fallbackDialer := &tls.Dialer{Config: &fallbackTLSConfig}
+		fbConn, fbErr := fallbackDialer.DialContext(ctx, "tcp", fallbackServerAddr)
+		if fbErr != nil {
+			grpclog.Infof("dialing to fallback server %s failed: %v", fallbackServerAddr, fbErr)
+			return nil, nil, fmt.Errorf("dialing to fallback server %s failed: %v; S2A client handshake with %s error: %w", fallbackServerAddr, fbErr, originServer, originErr)
+		}
+
+		tc, success := fbConn.(*tls.Conn)
+		if !success {
+			grpclog.Infof("the connection with fallback server is expected to be tls but isn't")
+			return nil, nil, fmt.Errorf("the connection with fallback server is expected to be tls but isn't; S2A client handshake with %s error: %w", originServer, originErr)
+		}
+
+		tlsInfo := credentials.TLSInfo{
+			State: tc.ConnectionState(),
+			CommonAuthInfo: credentials.CommonAuthInfo{
+				SecurityLevel: credentials.PrivacyAndIntegrity,
+			},
+		}
+		if grpclog.V(1) {
+			grpclog.Infof("ConnectionState.NegotiatedProtocol: %v", tc.ConnectionState().NegotiatedProtocol)
+			grpclog.Infof("ConnectionState.HandshakeComplete: %v", tc.ConnectionState().HandshakeComplete)
+			grpclog.Infof("ConnectionState.ServerName: %v", tc.ConnectionState().ServerName)
+		}
+		originConn.Close()
+		return fbConn, tlsInfo, nil
+	}, nil
+}
+
+// DefaultFallbackDialerAndAddress returns a TLS dialer and a network address for it to dial with.
+// A server name is derived from the `fallbackAddr`, and is used to initialize the TLS config of the fallback dialer.
+// example use:
+//
+//	    fallbackDialer, fallbackServerAddr := fallback.DefaultFallbackDialerAndAddress(fallbackAddr)
+//		dialTLSContext := s2a.NewS2aDialTLSContextFunc(&s2a.ClientOptions{
+//			S2AAddress:         s2aAddress, // required
+//			EnableV2:           true, // must be true
+//			FallbackOpts: &s2a.FallbackOptions{
+//				FallbackDialer: &s2a.FallbackDialer{
+//					Dialer:     fallbackDialer,
+//					ServerAddr: fallbackServerAddr,
+//				},
+//			},
+//	})
+//
+// The fallbackAddr is expected to be a network address, e.g. example.com:port. If port is not specified,
+// it uses default port 443.
+func DefaultFallbackDialerAndAddress(fallbackAddr string) (*tls.Dialer, string, error) {
+	var fallbackDialer *tls.Dialer
+	fallbackServerAddr, fallbackServerName, err := processFallbackAddr(fallbackAddr)
+	if err != nil {
+		return nil, "", err
+	}
+	fallbackTLSConfig := tls.Config{
+		ServerName: fallbackServerName,
+	}
+	fallbackDialer = &tls.Dialer{Config: &fallbackTLSConfig}
+
+	return fallbackDialer, fallbackServerAddr, nil
+}
+
+func processFallbackAddr(fallbackAddr string) (string, string, error) {
+	// hostname:port
+	var fallbackServerAddr string
+	// hostname
+	var fallbackServerName string
+	var err error
+
+	if fallbackAddr != "" {
```

Comment:
nit: Can we reverse the logic here so that we first check if fallbackAddr is the empty string and if so return early, and then have the rest of the logic?